### PR TITLE
add UIPulldownMenu widget

### PR DESCRIPTION
**Pull request by JimMarlowe**
This PR adds a UIWidget for a simplified pulldown menu, that one doesn't has to be a rocket scientist :rocket:  in order to create and use. With this, you can make a menubar enabled program with ease.
It also exposes two embedded functions for Typescript usage, onChanged() and onEvent().  Others (@Nyt0x, un PR'd) have defined onChanged as `onChanged: (data:any) => void;`, but I don't see anywhere actually filling in the event structure, so I left it out.

## UIPulldownMenu
This is a simplified Pulldown menu widget for user applications. It can be programmed by a .ui.txt layout file, or thru api, with a UISelectItemSource. It is designed to use the EVENT_TYPE_CHANGED event to inform when a menu entry has been selected. This event can be utilized in Javascript/Typescript code with the onChanged() function, and with C#,C++,  subscribe to the WidgetEvent, and check for EVENT_TYPE_CHANGED event type. You don't get the selected entry directly (for many reasons), so you have to ask the UIPulldownMenu which entry(ID) caused the CHANGED event. And, remember to fill out the id fields on the menu entries, or there won't be any ids to retrieve.

JJ/TS snippet
```
// in JS/TS
var filemenu = layout.getWidget("FileMenu");
filemenu.onChanged = function () 
{
    Atomic.print ( "FileMenu selected = " + filemenu.getSelectedId() ); 
    if ( filemenu.getSelectedId() === "FileExit" ) Atomic.engine.exit();
};

```
C# snippet
```
// and with C#
AtomicEngine.UIPulldownMenu filemenu = (AtomicEngine.UIPulldownMenu)layout.GetWidget("FileMenu");
filemenu.SubscribeToEvent<WidgetEvent>(filemenu, ev =>
{
    if (ev.Type == UI_EVENT_TYPE.UI_EVENT_TYPE_CHANGED)
    {
        string myitem = filemenu.GetSelectedId();
        Log.Info("FileMenu selected = " + myitem);
        if ( myitem == "FileExit" ) GetSubsystem<Engine>().Exit();
    }
});
```

sample_layout.ui.txt file
```

# UIPulldownMenu demo layout file
TBLayout: axis: y, size: available, distribution: available
	TBLayout: axis: x, id: Menubar, margin: 5, size: preferred, distribution: gravity, distribution-position: left right, skin: TBWindow
		TBPulldownMenu: text: "File", skin: 0, id: "FileMenu"
			items:
				item: text: "Open...", id: FileOpen
				item: text: "Save...", id: FileSave
				item: text: "-"
				item: text: "Exit", id: FileExit
		TBPulldownMenu: text: "Edit", skin: 0, id: "EditMenu"
			items:
				item: text: "Cut", id: EditCut
				item: text: "Copy", id: EditCopy
				item: text: "Paste", id: EditPaste
				item: text: "Delete", id: EditDelete
				item: text: "-"
				item: text: "Select All", id: EditAll
		TBPulldownMenu: text: "View", skin: 0, id: "ViewMenu"
			items:
				item: text: "Scale", id: ViewScale
				item: text: "Rotate", id: ViewRotate
				item: text: "Translate", id: ViewTranslate
		TBPulldownMenu: text: "Help", skin: 0, id: "HelpMenu"
			items:
				item: text: "Help...", id: HelpHelp
				item: text: "About this program...", id: HelpAbout
	TBWidget
	TBWidget
	TBWidget
	TBWidget
	TBButton: text: Press me!

```

### CODE REVIEW

File: Script/Packages/Atomic/UI.json [hunk 1]
```diff
@@ -22,7 +22,9 @@
 		    "getWidgetAt<T extends UIWidget>(x: number, y: number, include_children: boolean): T;"
 		],
 		"UIWidget": [
-		    "getWidget<T extends UIWidget>(id: string): T;"
+		    "getWidget<T extends UIWidget>(id: string): T;",
+		    "onEvent: () => void;",
```
Comment (JoshEngebretson) [inline Script/Packages/Atomic/UI.json]: Good getting these added to the TS defs, ```onEvent``` takes an Atomic.UIWidgetEvent parameter: https://github.com/AtomicGameEngine/AtomicGameEngine/blob/master/Source/AtomicJS/Javascript/JSUI.cpp#L273, you can find the TypeScript decl in the generated Atomic.d.ts file for reference

```typescript
** object returned in the callback for the WidgetEvent event.**/
    export interface UIWidgetEvent extends Atomic.EventData {
        /**  pointer of widget's OnEvent we are in */
        handler : Atomic.UIWidget;
        target : Atomic.UIWidget;
        type : Atomic.UI_EVENT_TYPE;
        x : number;
        y : number;
        deltaX : number;
        deltaY : number;
        count : number;
        key : number;
        specialKey : number;
        modifierKeys : number;
        refid : string;
        touch : boolean;
        /**  -> bool */
        /** Unmapped Native Type: [OUT] */
        handled : any;
    }
```

File: Source/Atomic/UI/UI.cpp [hunk 2]
```diff
@@ -741,6 +742,14 @@ UIWidget* UI::WrapWidget(tb::TBWidget* widget)
         return select;
     }
 
+    if (widget->IsOfType<TBPulldownMenu>())
+    {
+        UIPulldownMenu* select = new UIPulldownMenu(context_, false);
+        select->SetWidget(widget);
+        WrapWidget(select, widget);
+        return select;
+    }
```
Comment (JoshEngebretson) [inline Source/Atomic/UI/UI.cpp]: I wonder if we can automate this at some point, Atomic UI magic in the meantime 🐇 🎩

File: Source/ThirdParty/TurboBadger/tb_atomic_widgets.cpp [hunk 3]
```diff
@@ -638,5 +640,124 @@ TBWidget *TBFinderWindow::FindParentList( TBWidget *widget) // utility for deali
...
+bool TBPulldownMenu::OnEvent(const TBWidgetEvent &ev)
+{
+    if ( ev.target->IsOfType<TBButton>() && ev.type == EVENT_TYPE_CLICK)
+    {
+        // Open the menu, or set the value and close it if already open (this will
+        // happen when clicking by keyboard since that will call click on this button)
+        if (TBMenuWindow *menu_window = GetMenuIfOpen())
+        {
+            TBWidgetSafePointer tmp(this);
+            int value = menu_window->GetList()->GetValue();
```
Comment (JoshEngebretson) [inline Source/ThirdParty/TurboBadger/tb_atomic_widgets.cpp]: Say what you will, though you're getting pretty darn good with TB :)

File: Source/ThirdParty/TurboBadger/tb_atomic_widgets.cpp [hunk 4]
```diff
@@ -638,5 +640,124 @@ TBWidget *TBFinderWindow::FindParentList( TBWidget *widget) // utility for deali
...
+            SetValue(menu_window->GetList()->GetValue());
+        return true;
+    }
+    else if (ev.target == this && m_source && ev.IsKeyEvent())
+    {
+        if (TBMenuWindow *menu_window = GetMenuIfOpen())
+        {
+            // Redirect the key strokes to the list
+            TBWidgetEvent redirected_ev(ev);
+            return menu_window->GetList()->InvokeEvent(redirected_ev);
```
Comment (JoshEngebretson) [inline Source/ThirdParty/TurboBadger/tb_atomic_widgets.cpp]: Whoa, serious code-fu!

agent-smith-neo-in-the-matrix-revolutions-agent-smith-22575856-570-300

File: Script/Packages/Atomic/UI.json (hunk 1 above)
Comment (JimMarlowe) [inline Script/Packages/Atomic/UI.json]: doh. I should have seen the  PushWidgetEventObject(eventData); I'll fix that.

File: Source/Atomic/UI/UI.cpp (hunk 2 above)
Comment (JimMarlowe) [inline Source/Atomic/UI/UI.cpp]: UI magic is fine for the amount of traffic it gets.

**Comment by JimMarlowe** [comment]
This is all leading up to a TB periodic table of widgets sample program, that shows what widgets you can use, and how to get at them from your program in JS/C#/C++. I keep pecking :baby_chick:  away at it, but TB usability leaves a lot to be desired. I am often amazed at what you've gone thru with the editor, hence, this widget.

File: Script/Packages/Atomic/UI.json [hunk 5]
```diff
@@ -22,7 +22,9 @@
 		    "getWidgetAt<T extends UIWidget>(x: number, y: number, include_children: boolean): T;"
 		],
 		"UIWidget": [
-		    "getWidget<T extends UIWidget>(id: string): T;"
+		    "getWidget<T extends UIWidget>(id: string): T;",
+		    "onEvent: (data:any) => void;",
```
Comment (JoshEngebretson) [inline Script/Packages/Atomic/UI.json]: Sorry if wasn't clear, I think we know this is a UIWidgetEvent, so ```eventData:UIWidgetEvent``` should work here as a parameter?

**Comment by JoshEngebretson** [comment]
I look forward to the UI periodic table :) "I am often amazed at what you've gone thru with the editor", that is exactly right... part of the evolution of Atomic, this would have made things better on the frontend :)

**Comment by JimMarlowe** [comment]
Changed and tested (it's good).

**Comment by JoshEngebretson** [comment]
Great, thanks! LANDING! ✈️

**Pull request merged**